[Compose 페이먼츠 1단계] 동전 미션 제출합니다.

donghyun81 opened the pull request:
# 셀프 체크리스트

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

- Ui 테스트에서 계속 오류가 발생해서 테스트는 작성하지 못했습니다.. 추후 작성하고 다시 PR 요청하겠습니다!
- 재사용 가능한 화면의 기준을 잡지 못해서 관련한 내용 리뷰 해주시면 좋을 것 같습니다!
- 각 화면의 위치와 거리 조절을 어디서 해야할지 감을 잡기 어려워서 조합되는 화면 단위로 작성했는데, 의논해보면 좋을 것 같습니다
- 각 상태와 이벤트에 관해서 의논해보면 좋을 것 같습니다.
- VisualTransformation은 Compose 학습과 거리가 있다고 생각해서 생성형 AI를 통해서 구현했습니다!

## 코드 리뷰 커뮤니케이션

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

https://github.com/user-attachments/assets/ea61beed-e1f4-41cd-9d89-c1deb3146c40



![]()

donghyun81 committed:
docs: 기능 구현 목록 정리

donghyun81 committed:
feat: 카드 추가 화면 UI 구현

donghyun81 committed:
feat: 카드 추가 화면 UI 구현

donghyun81 committed:
build: ktlint 적용 및 코드 포맷팅

wondroid-world commented on app/src/main/java/woowacourse/payments/MainActivity.kt:
```diff
@@ -20,28 +18,13 @@ class MainActivity : ComponentActivity() {
-                        name = "Android",
-                        modifier = Modifier.padding(innerPadding)
+                    CreateCardScreen(
+                        { TODO() },
```
현재 `CreateCardScreen`에 람다가 두개가 있네용!
각 람다가 어떤 인자에 해당되는 람다인지 헷갈리는 데, 동전은 어떻게 생각하실까용?

donghyun81 replied:
[refactor: NewCardTopBar의 매개변수 이름을 명시적으로 변경](https://github.com/woowacourse/android-payments/pull/24/commits/aa491d3bb8c8b78808e3154df56fa61c607533b0)
기초적인걸 놓치고 있었네요 named arguments를 통해서 개선 했습니다!

wondroid-world commented on app/src/main/java/woowacourse/payments/domain/ExpiryException.kt:
```diff
@@ -0,0 +1,13 @@
+package woowacourse.payments.domain
+
+sealed class ExpiryException : Throwable() {
```
오! sealed class로 에러 타입을 명시해줘서 너무 좋아요! 👍

wondroid-world replied:
에러명을 `ExpiryException`로 명시한 이유가 있을까용?

donghyun81 replied:
[refactor: 카드 만료일 유효성 검증 로직 변경](https://github.com/woowacourse/android-payments/pull/24/commits/45cfc5dde030079c24d0a5f544698156ae68de1c)
추후에 도메인의 예외로 사용될 것 같아서 ExpiryException으로 명시했는데 카드 만료일이라고 수정하는게 좋을 것 같아서 변경했습니다!

donghyun81 replied:
[refactor: 카드 만료일 유효성 검증 로직 변경](https://github.com/woowacourse/android-payments/pull/24/commits/45cfc5dde030079c24d0a5f544698156ae68de1c)

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/cardcreate/CreateCardInputSection.kt:
```diff
@@ -0,0 +1,104 @@
+            onValueChange = { password ->
+                onCardChange(createCardState.copy(password = password.take(PASSWORD_MAX)))
+            },
+            keyboardOptions = KO_PIN,
```
각 상황에 맞는 키보드 명시 너무 좋네요! 💯

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/cardcreate/CreateCardInputSection.kt:
```diff
@@ -0,0 +1,104 @@
+    createCardErrorState: CreateCardErrorState,
+    onCardChange: (CreateCardState) -> Unit,
+    onErrorChange: (CreateCardErrorState) -> Unit,
+    modifier: Modifier,
```
동전! 레아가 주신 링크를 보시면, 
가이드 라인에 UI를 내는 **컴포저블은 modifier: Modifier = Modifier를 단 하나 두고, 첫 번째 선택적 파라미터로 둡니다.**라는 말이 있었습니당! 이유는 "호출부에서 modifier를 포지셔널 인자로 쉽게 줄 수 있고, 일관적입니다."라고 나와있습니당! 아래 사이트 참고 부탁드립니당! 

[API Guidelines for Jetpack Compose](https://android.googlesource.com/platform/frameworks/support/+/androidx-main/compose/docs/compose-api-guidelines.md)

wondroid-world replied:
다른 곳에서 `default`값으로 `Modifier`을 넣어 주셨던 데, 여기서는 안 넣으신 이유가 있으실까용? 
동전의 의도가 궁금해용!

donghyun81 replied:
[refactor: 카드 추가 화면 상태 관리 로직 개선](https://github.com/woowacourse/android-payments/pull/24/commits/d45ed97779d290c055b96171146dbf698be8af00)
어떤 방법이 맞는지 모르는 상태로 이것저것 시도해보다가 정리를 못했던 것 같습니다.. ㅎㅎ 말씀해주신 방법으로 변경했습니다!

wondroid-world replied:
P3; 동전!
`modifier`에 기본값 있으면 장점이 호출 코드가 간결해진다는 장점이 있다고 생각해용! 대부분의 Compose 라이브러리 컴포넌트가 이 방식을 사용하고 있고용! 동전이 생각하시기에 `modifier`에 default 값이 있을 때와 없을 때의 차이는 어떤거라고 생각하실까용? ❓

donghyun81 replied:
? 이거 defalut 값이 없네요???
저는 modifier의 default 값이 있으면 호출부에서 필요에 의해서만 해당 컴포넌트의 위치나 크기를 조정할 수 있다는 장점이 있는 것 같습니다! 애초에 optional parameter로 위치를 제일 아래에 선언했는데 defalut 값이 없네요 ㅎㅎ..
2단계에서 반영하겠습니다!

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/cardcreate/CreateCardInputSection.kt:
```diff
@@ -0,0 +1,104 @@
+    ) {
+        CreateCardNumbersInput(
+            value = createCardState.cardNumber,
+            labelText = stringResource(R.string.card_number_label),
```
상수화 좋네용!

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/cardcreate/CreateCardInputSection.kt:
```diff
@@ -0,0 +1,104 @@
+            onValueChange = { numbers ->
+                onCardChange(createCardState.copy(cardNumber = numbers))
+            },
+            visualTransformation = cardVt,
```
`cardVt`는 어떤 의미일까용?

donghyun81 replied:
[refactor: GroupedSeparatorVisualTransformation 인스턴스를 remember로 감싸도록 변경](https://github.com/woowacourse/android-payments/pull/24/commits/cd9e3d142603396095a6912e88ace0d08f9ed9c2)
cardNumbers에 관한 VisualTransformation이였습니다! 제가 봐도 이해하기 힘들 것 같아서 수정했습니다!

wondroid-world commented on app/src/main/res/values/colors.xml:
```diff
@@ -7,4 +7,6 @@
     <color name="black">#FF000000</color>
     <color name="white">#FFFFFFFF</color>
+    <color name="gray_33">#333333</color>
+    <color name="yellow_CB">#CBBA64</color>
```
etpack Compose에서는 Kotlin 코드 안에서 색을 정의하기 때문에, 가이드라인이 달라진다고 해요! 이것도 참고하면 좋을거 같아요!

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/cardcreate/CreateCardScreen.kt:
```diff
@@ -0,0 +1,150 @@
+fun CreateCardScreen(
+    onBackClick: () -> Unit,
+    onSaveClick: () -> Unit,
+    modifier: Modifier = Modifier,
```
이 부분도 가이드라인에 따라서 위치를 변경해보면 어떨까용?

donghyun81 replied:
해당 부분은 첫 번째 optional parameter에 modifier가 있는 것 같은데 어떤 위치로 이동하는걸 말씀하신걸까요??

wondroid-world replied:
아! 이부분은 제가 실수했습니다! 혼란을 드려 죄송해용 동전 😢

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/cardcreate/CreateCardScreen.kt:
```diff
@@ -0,0 +1,150 @@
+import woowacourse.payments.ui.cardcreate.model.CreateCardErrorState
+import woowacourse.payments.ui.cardcreate.model.CreateCardState
+
+private val CardWidth = 208.dp
```
상수 요소 들이 증가하면 관리하기 한 파일의 라인 수가 많이 증가할거 같은 데, 이 상수들을 어떻게 관리하면 좋을까용?

donghyun81 replied:
dimen과 로컬에서 고민하다가 공통 분모를 찾기 어려워서 로컬 파일 변수로 두었는데,
NowInAndorid에서는 각 공통요소가 불분명하면 각 컴포넌트에 직접 사용하는 방식을 사용하고 있어서 해당 방식으로 수정했습니다!

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/components/TextField.kt:
```diff
@@ -0,0 +1,38 @@
+import androidx.compose.ui.text.input.VisualTransformation
+
+@Composable
+fun PaymentsOutlinedTextField(
```
공통 UI를 만들었네용! 굿굿!!

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/cardcreate/CreateCardInputSection.kt:
```diff
@@ -0,0 +1,104 @@
+    onErrorChange: (CreateCardErrorState) -> Unit,
+    modifier: Modifier,
+) {
+    val cardVt = remember { GroupedSeparatorVisualTransformation(CARD_GROUPS, SEP_CARD) }
```
`CreateCardInputSection`에 들어갈 컴포넌트들의 상태를 모두 위로 끌어올려주셨군요 !
컴포넌트들의 상태는 각 컴포넌트 안에 배치 할 수도 있었을 것 같은데
이렇게 구현하게된 동전의 이유가 궁금합니다 

> 추가로 레아께서 페토 리뷰에 남기신 내용인데 동전에게도 생각의 도움이 될거같아 리뷰를 남깁니당!
> 
> 모두 같아 보이는데 어떤 차이가 있을까요?
> 
> var username by remember { mutableStateOf("") }
> val username = remember { mutableStateOf("") }
> val (usernameValue, setUsername) = remember { mutableStateOf("") }

donghyun81 replied:
해당 구현에서 상태 호이스팅을 한 이유는 
1. 인자값을 외부에서 받으면 각 컴포넌트를 테스트시에 원하는 값을 인자로 사용할 수 있습니다.
2. 하나의 단일 소스이기 때문에 각 상태를 다른 컴포넌트와 공유할 수 있습니다.
    - 만료일에서 검증 문제가 생겼을 경우에 에러 상태를 변경 시킬 수 있습니다.
 
 좋은 질문 주셔서 알아본 내용은

>  var username by remember { mutableStateOf("") } -> mutableStateOf의 get,set을 by로 위임해서 username을 그대로 사용가능합니다.
> val username = remember { mutableStateOf("") } -> .value를 통해 값을 조회하거나 수정해야합니다.
> val (usernameValue, setUsername) = remember { mutableStateOf("") } ->
> usernameValue는 반환이 String으로 getter의 역할을 하고 setUsername는 (String) -> Unit 콜백으로 setter의 역할을 하게 됩니다.

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/utils/GroupedSeparatorVisualTransformation.kt:
```diff
@@ -0,0 +1,60 @@
+import androidx.compose.ui.text.input.TransformedText
+import androidx.compose.ui.text.input.VisualTransformation
+
+class GroupedSeparatorVisualTransformation(
```
오! `GroupedSeparatorVisualTransformation`를 객체로 분리해 주셨네용!
익명 객체로 만들 수도 있었는 데, 객체로 만든 동전의 이유가 궁금해용!

donghyun81 replied:
GroupedSeparatorVisualTransformation은 그룹들의 개수들과 구분자를 인자로 받아서 생성하는VisualTransformation타입의 객체라서 인자를 받아서 재사용하기 위해서 객체로 분리했습니다! 익명 객체라면 매번 재구현해야 할 것 같아요!

wondroid-world reviewed (commented):
안녕하세요! 동전! 
매일 같은 팀으로 만나는 데, 이렇게 리뷰어로써 만나게 되다니 영광입니당! 

동전의 코드를 보면서 궁금한 점을 남기게 되었습니당! 
언제든지 열려있으니, 같이 컴포즈에 대해서 공부하고 의견을 나누면 좋을거 같아요! 😄 

오늘도 화이팅!! 👍

woowahan-leah commented on app/src/main/java/woowacourse/payments/ui/cardcreate/CreateCardScreen.kt:
@donghyun81 왜 테스트 안짜세요 😉 
@wondroid-world 왜 테스트 리뷰 안남겨주세요 😉

donghyun81 replied:
오늘 수업에서 배운 내용 관련해서 계속 테스트에 실패해서 계속 잡고있다가, 테스트 작성을 못했습니다.. ㅎㅎ

wondroid-world reviewed (changes requested):
동전 수고하셨습니당! 리뷰 확인부탁드립니당 🪙 👍

donghyun81 committed:
refactor: 카드 만료일 유효성 검증 로직 변경

- `ExpiryException`을 `CardExpiryException`으로 이름 변경
- `CardExpiryException`의 에러 메시지를 `strings.xml`에서 관리하도록 변경
- 만료일 형식 검증 로직 개선

donghyun81 committed:
refactor: 카드 추가 화면 상태 관리 로직 개선

- 카드 추가 화면의 입력 필드 상태 관리를 `CreateCardStateHolder`로 이전
- `CreateCardInputs.kt` 와 `TextField.kt` 파일 삭제 및 관련 로직 `CreateCardInputSection.kt` 로 통합
- `CreateCardState` 에 유효기간 오류 메시지 리소스 ID 추가
- `CreateCardErrorState` 데이터 클래스 삭제

donghyun81 committed:
feat: 카드 추가 화면 TopBar를 Scaffold TopBar 속성으로 이전

donghyun81 committed:
refactor: `NewCardTopBar`의 매개변수 이름을 명시적으로 변경

donghyun81 committed:
refactor: PaymentCard 컴포저블 분리

donghyun81 committed:
refactor: `GroupedSeparatorVisualTransformation` 인스턴스를 remember로 감싸도록 변경

donghyun81 committed:
refactor: `colorResource`를 제거하고 `Color` 객체를 사용하도록 수정

donghyun81 committed:
test: 카드 추가 화면 입력 필드 UI 테스트 추가

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/utils/ext/StringExt.kt:
```diff
@@ -1,19 +1,19 @@
-    val yy = digits.substring(2, 4).toIntOrNull() ?: return ExpiryException.InvalidYear
-    if (mm !in 1..12) return ExpiryException.InvalidMonth
+    if (digits.length != 4) return null
+    val mm = digits.substring(0, 2).toIntOrNull() ?: return CardExpiryException.InvalidMonth
```
P5; 동전 이건 메다 센세가 알려준건데,
`take`를 사용해서 `digits.take(2).toIntOrNull`이라고 표현할 수도 있다고 해용!
`take`가 코틀린에서 생긴 API이라서 좀더 코틀린 스럽게 작성하기 위해서, `take`도 있다는 것을 고려부탁드려용!

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/utils/ext/StringExt.kt:
```diff
@@ -1,19 +1,19 @@
     val now = YearMonth.now()
     val exp = YearMonth.of(2000 + yy, mm)
-    if (exp.isBefore(now)) return ExpiryException.Expired
+    if (exp.isBefore(now)) return CardExpiryException.Expired
```
P4; `exp`를 보니, 오늘 이브이가 생각나네용! 조금더 서술적인 변수명도 좋을거 같아요!

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/utils/ext/StringExt.kt:
```diff
@@ -1,19 +1,19 @@
+import woowacourse.payments.domain.CardExpiryException
 import java.time.YearMonth
 
 // 추후 도메인으로 이전
```
도메인 귯!!

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/cardcreate/CreateCardStateHolder.kt:
```diff
@@ -0,0 +1,53 @@
+import woowacourse.payments.ui.utils.ext.formatCardExpiryException
+import woowacourse.payments.ui.utils.ext.toErrorResourceId
+
+class CreateCardStateHolderSaver : Saver<CreateCardStateHolder, CreateCardState> {
```
P5; `Saver<CreateCardStateHolder, CreateCardState>`는 뭘까용? 처음 보는 거라서 궁금해용!!

wondroid-world commented on app/src/main/java/woowacourse/payments/MainActivity.kt:
```diff
@@ -17,10 +18,10 @@ class MainActivity : ComponentActivity() {
-                Scaffold(modifier = Modifier.fillMaxSize()) { innerPadding ->
+                Scaffold(
+                    modifier = Modifier.fillMaxSize(),
+                    topBar = { NewCardTopBar({}, {}) }) { innerPadding ->
```
`innerPadding`의 역할이 뭔지 궁금해용!

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/theme/Color.kt:
```diff
@@ -9,3 +9,6 @@ val Pink80 = Color(0xFFEFB8C8)
 val PurpleGrey40 = Color(0xFF625b71)
 val Pink40 = Color(0xFF7D5260)
+
+val Gray33 = Color(0xFF333333)
```
굿굿!!

wondroid-world commented on app/src/main/java/woowacourse/payments/ui/cardcreate/CreateCardInputSection.kt:
```diff
@@ -50,7 +56,9 @@ fun CreateCardInputSection(
-            modifier = Modifier.fillMaxWidth(),
+            modifier = Modifier
+                .fillMaxWidth()
+                .testTag(CARD_NUMBERS_INPUT_TAG),
```
P5; 동전은 production code에 test관련 코드가 들어가는 점에 대해서 어떻게 생각하는 지 궁금해용!!

wondroid-world commented:
<img width="344" height="736" alt="image" src="https://github.com/user-attachments/assets/e3d3797f-8ff3-42ab-a668-5d6d3f4fa063" />

비밀번호 입력할 때, 화면이 올라가는 점 확인부탁드립니당!!!

wondroid-world reviewed (approved):
동전 짧은 시간내에 구현하신다고 수고많으셨습니다!!
역시 동전이 동전했군용! 

미션이랑 팀플 모두다 화이팅입니당~ 👍
